runtime/include/spinlock.hpp

```diff
@@ -0,0 +1,59 @@
+#ifndef SPINLOCK_HPP
+#define SPINLOCK_HPP
+
+#include <atomic>
+#include <cstddef>
+#include <ctime>
+#include <time.h>
+
+class Spinlock {
+    std::atomic<unsigned int> lock;
+public:
+
+    // Default constructor to initialize the atomic_flag
+    Spinlock() : lock(0) {}
+
+    // Deleted copy constructor to prevent copying
+    Spinlock(const Spinlock&) = delete;
+    Spinlock& operator=(const Spinlock&) = delete;
+
+    Spinlock(Spinlock&& other) noexcept : lock() {
+        if (other.lock.exchange(1, std::memory_order_acquire)) {
+            lock.store(0, std::memory_order_release);
+        }
+    }
+
+    Spinlock& operator=(Spinlock&& other) noexcept {
+        if (this != &other) {
+            if (other.lock.exchange(1, std::memory_order_acquire)) {
+                lock.store(0, std::memory_order_release);
+            }
+        }
+        return *this;
+    }
+
+    void spin_lock() {
+        static const timespec ns = {0, 1};
+        for(int i = 0; lock.load(std::memory_order_relaxed) || lock.exchange(1, std::memory_order_acquire); ++i) {
+            if(i==8){
+                i=0;
+                nanosleep(&ns, NULL);
```

Comment:
Nanosleep is a syscall. It would be better if there is a pure user space spinlock implementation. Syscall are very slow and we should avoid them in user space eBPF.

If there is any similar instruction like pause on x86, replace the loop body with it using inline assembly. If not, just leave an empty loop body

Reply:
For future reference, keep in mind that compilers re-order instructions at will, that's why it is common to have a memory barrier in lock implementations.

I see you are using the relaxed memory order, but I am not confident it will do the trick. Keep it in mind when/if implementing the in-line asm version as well. I am particularly not sure if/how compilers also move in-line ASM around.

https://en.cppreference.com/w/cpp/atomic/memory_order

Reply:
Noted @Officeyutong 


> For future reference, keep in mind that compilers re-order instructions at will, that's why it is common to have a memory barrier in lock implementations.

Thanks for telling me about this. 

> I see you are using the relaxed memory order, but I am not confident it will do the trick. Keep it in mind when/if implementing the in-line asm version as well. I am particularly not sure if/how compilers also move in-line ASM around.

I have seen it in many blogs being used, so I am guessing it should work, but yea I will look into it, as for in-line asm, will check, what can I replace it with, Thanks for your input.

> https://en.cppreference.com/w/cpp/atomic/memory_order
Thanks for the reference

Reply:
This https://github.com/gstrauss/plasma/blob/3d5b64539f6aba1218e303be7e33f47c6ad74c22/plasma_spin.h#L94 
has used `yield`